### ISSUE TITLE
the unix listener closes the connection unexpectedly with: echo foo | nc -U /tmp/mysock

### ISSUE THREAD

**Issue by Phibonacci** [agronholm/anyio#200]
I tried the following example on the [AnyIO documentation](https://anyio.readthedocs.io/en/stable/networking.html#working-with-unix-sockets):

```python
from anyio import create_unix_listener, run


async def handle(client):
    async with client:
        name = await client.receive(4) # I changed the value
        await client.send(b'Hello, %s\n' % name)


async def main():
    listener = await create_unix_listener('/tmp/mysock')
    await listener.serve(handle)

run(main)
```

When sending a string with `echo foo | nc -U /tmp/mysock` the answer is never printed. However running `nc -U /tmp/mysock` interactively it is working.

Also this callback works as expected with `asyncio.start_unix_server()`:

```python
    async def _handle_request(self, reader, writer):
        line = await reader.readline()
        writer.write(line)
        await writer.drain()
        writer.write_eof()
        writer.close()
```

**Comment by joernheissler** [comment]
Code works with trio backend, doesn't with asyncio.

strace shows:
```c
accept4(6, {sa_family=AF_UNIX}, [110->2], SOCK_CLOEXEC) = 7
getsockname(7, {sa_family=AF_UNIX, sun_path="/tmp/mysock"}, [128->14]) = 0
ioctl(7, FIONBIO, [1])                  = 0
epoll_wait(3, [], 2, 0)                 = 0
epoll_ctl(3, EPOLL_CTL_DEL, 6, 0x7fff40d06cc4) = 0
ioctl(7, FIONBIO, [1])                  = 0
getsockname(7, {sa_family=AF_UNIX, sun_path="/tmp/mysock"}, [110->14]) = 0
getpeername(7, {sa_family=AF_UNIX}, [110->2]) = 0
epoll_wait(3, [], 1, 0)                 = 0
getpid()                                = 495823
getpid()                                = 495823
epoll_ctl(3, EPOLL_CTL_ADD, 7, {EPOLLIN, {u32=7, u64=7}}) = 0
epoll_wait(3, [{EPOLLIN, {u32=7, u64=7}}], 2, 0) = 1
getpid()                                = 495823
getpid()                                = 495823
getpid()                                = 495823
recvfrom(7, "foo\n", 262144, 0, NULL, NULL) = 4
epoll_wait(3, [{EPOLLIN, {u32=7, u64=7}}], 2, 0) = 1
getpid()                                = 495823
getpid()                                = 495823
accept4(6, 0x7fff40d06b10, [110], SOCK_CLOEXEC) = -1 EAGAIN (Resource temporarily unavailable)
epoll_ctl(3, EPOLL_CTL_ADD, 6, {EPOLLIN, {u32=6, u64=6}}) = 0
recvfrom(7, "", 262144, 0, NULL, NULL)  = 0
epoll_ctl(3, EPOLL_CTL_DEL, 7, 0x7fff40d06af4) = 0
epoll_wait(3, [], 2, 0)                 = 0
getpid()                                = 495823
getpid()                                = 495823
close(7)                                = 0
```

**Comment by agronholm** [comment]
The test suite tests for EOF handling but apparently not thoroughly enough.

**Comment by agronholm** [comment]
I've created a test based on that example that fails only on asyncio.

**Comment by agronholm** [comment]
As I suspected, the test also fails with TCP since they share most of the socket stream code.

**Comment by agronholm** [comment]
Curiously, it doesn't fail on plain asyncio + TCP, only uvloop + TCP.

**Comment by agronholm** [comment]
I factored out the anyio client side by switching to nonblocking stdlib sockets and now asyncio is failing all the new tests while curio and trio pass them.

**Comment by agronholm** [comment]
The connection is closed from the server end right after the handler task is spawned. Even if I do nothing but sleep in the handler task, the client receives notice of the connection closing. I'm trying to figure out why this is happening.

**Comment by agronholm** [comment]
I managed to reproduce this without any AnyIO code.

Server:
```python3
import socket
from asyncio import Protocol, get_event_loop, sleep


async def main():
    client, _ = await loop.sock_accept(serv_sock)
    transport, protocol = await loop.connect_accepted_socket(Protocol, client)
    await sleep(1)
    transport.write(b'Hello\n')
    transport.close()


serv_sock = socket.socket()
serv_sock.bind(('127.0.0.1', 5555))
serv_sock.listen()
serv_sock.setblocking(False)
loop = get_event_loop()
loop.run_until_complete(main())
```
Client:
```python3
import socket

s = socket.socket()
s.connect(('127.0.0.1', 5555))
s.shutdown(socket.SHUT_WR)
print('received:', s.recv(10))
s.close()
```
The connection is closed by the server during that `await sleep(1)`.

**Comment by agronholm** [comment]
Ok, I figured it out. There's something I didn't know – you have to return `True` from a Protocol's `eof_received()` method, or else the transport will be shut down.

**Comment by agronholm** [comment]
The fix makes the tests pass but then the TLSListener test fails, but only on plain asyncio, not uvloop. Sigh.

**Comment by agronholm** [comment]
Fixed via 3b4c6495, released in v2.2.0.
